src/richie/apps/search/indexers/courses.py

```diff
@@ -87,310 +217,193 @@ class CoursesIndexer:
+        # --------------------- weight_2 ---------------------
         #      Bucket 2: Courses that are archived but open
-        #       sorted by descending end of enrollment date
+        #       sorted by ascending end of enrollment date
         #
-        # --------------   3 x datetime.MAXYEAR  --------------
+        # --------------------- weight_3 ---------------------
         #  Bucket 3: Courses that are future but not yet open
         #            sorted by ascending start date
         #
-        # --------------   4 x datetime.MAXYEAR  --------------
+        # --------------------- weight_4 ---------------------
         # Bucket 4: Courses that are future but already closed
-        #      sorted by descending end of enrollment date
+        #            sorted by ascending start date
         #
-        # --------------   5 x datetime.MAXYEAR  --------------
-        #     Bucket 5: Courses that are ongoing but closed
-        #      sorted by descending end of enrollment date
+        # --------------------- weight_5 ---------------------
+        #     Bucket 5: Courses that are on-going but closed
+        #           sorted by ascending end date
         #
-        # --------------   6 x datetime.MAXYEAR  --------------
+        # --------------------- weight_6 ---------------------
         #          Bucket 6: Courses that are archived
-        #             sorted by descending end date
+        #            sorted by descending end date
         #
-        # --------------   7 x datetime.MAXYEAR  --------------
-        #     Bucket 7: Courses that have no course runs
+        # ------------------ Weight set to 0 -----------------
+        #      Bucket 7: Courses that have no course runs
         #                No specific ordering
         #                   ———————————————
         #                   END OF THE LIST
         #
-        # For reference MAXYEAR's timestamp is more than 2 orders of magnitude larger than
-        # this year's timestamp (2018).
-        # This means there can be no overlap between the various buckets, but we can still
-        # sort courses inside each bucket as we see fit by simply adding timestamps (ascending
-        # order) or substracting them (descending order).
-        "state": {
+        "score": {
             "script": {
                 "lang": "painless",
-                "source": """
-                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern(
-                        "yyyy-MM-dd'T'HH:mm:ss[.SSSSSS]XXXXX"
+                "source": BEST_STATE_SCRIPT
+                + """
+                if (best_state == 0) {{
+                    // The course is on-going and open
+                    // Ordered by ascending end of enrollment datetime. The next course to
+                    // end enrollment is displayed first.
+                    return (_score + 1) * (
+                        {weight_0:d} * params.ms_since_epoch - ZonedDateTime.parse(
```

Comment:
Are there not missing parentheses here? It looks like we're subtracting the enrollment_end milliseconds from eg. 80x the milliseconds for now

Reply:
Yes! I understand it can look weird.
In fact:
- 80 * is to boost the score of this "bucket" 80 times compared to other buckets so that we have an impact on the relevance score and rank courses open for enrollment much better that courses that are in a worse state,
- substracting the enrollment_end date is just small ordering stimulus within the bucket so as soon as there is a fulltext search applied, ordering is no more determined by this but by the relevance score.

Reply:
I get it now, it's pretty clever but it does look like a mistake for the unaware observer. Maybe this deserves a comment so someone does not "fix" it later on 😅

Reply:
True! I will copy these 2 sentences to the docstring.